base/shared/src/main/scala/scalaz/data/io.scala

```diff
@@ -0,0 +1,109 @@
+package scalaz
+package data
+
+import scalaz.tc.Foldable
+import scalaz.zio.Errors.NothingRaced
+import scalaz.zio.{ Fiber, IO }
+
+import scala.{ AnyVal, Either, Nothing, Option, Unit }
+import Predef.identity
+import scala.collection.generic.CanBuildFrom
+import scala.collection.mutable
+import scala.inline
+
+object io {
+
+  // TODO check if this can be optimized
+  private final def _bracket0[E, A, B](
+    acquire: IO[E, A]
+  )(release: (A, Maybe[E \/ B]) => IO[Nothing, Unit])(use: A => IO[E, B]): IO[E, B] = {
+    val release1: (A, Option[Either[E, B]]) => IO[Nothing, Unit] =
+      (a: A, o: Option[Either[E, B]]) => release(a, Maybe.fromOption(o.map(Disjunction.fromEither[E, B])))
+    IO.bracket0[E, A, B](acquire)(release1)(use)
+  }
+
+  implicit class IOObjOps(private val ioObj: IO.type) extends AnyVal {
+
+    final def absolve[E, A](v: IO[E, E \/ A]): IO[E, A] =
+      v.flatMap(fromEither)
+
+    final def fromEither[E, A](v: E \/ A): IO[E, A] =
+      v.fold(IO.fail, IO.now)
+
+    @inline
+    final def bracket0[E, A, B](
+      acquire: IO[E, A]
+    )(release: (A, Maybe[E \/ B]) => IO[Nothing, Unit])(use: A => IO[E, B]): IO[E, B] =
+      _bracket0[E, A, B](acquire)(release)(use)
+
+    final def require[E, A](error: E): IO[E, Maybe[A]] => IO[E, A] =
+      (io: IO[E, Maybe[A]]) =>
+        io.flatMap { maybe =>
+          Maybe.fold[A, IO[E, A]](maybe)(IO.now[A], IO.fail[E](error))
+      }
+
+    // TODO replace CanBuildFrom with Unfoldable
+    final def forkAll[E, A, F[_]](
+      as: F[IO[E, A]]
+    )(implicit f: Foldable[F], cbf: CanBuildFrom[F[IO[E, A]], A, F[A]]): IO[E, Fiber[E, F[A]]] =
```

Comment:
Yeah let's replace this by `Unfoldable` as soon as your other PR is merged. 😄

Reply:
Should be pretty easy, since `Foldable` has a `toList` (which i'm gonna update returning an IList) and `Unfoldable` has a `fromList`.

But I'm afraid this approach will be inefficient (but elegant, of course).
What're your thoughts about it?

Reply:
It will definitely be inefficient. But we can deal with that later.

We need `Collection[F] extends Foldable[F] with Unfoldable[F]`, and then we can specialize on more common compositions.

Reply:
I actually run a couple of benchmark with the version i'm working on, and.. it's way too slow.
Will push that anyway if that's ok with you,, and optimize later on.